### ISSUE TITLE
pmseries accessing redis via pmproxy disconnects the client and triggers an assert

### ISSUE THREAD

**Issue by goodwinos** [performancecopilot/pcp#859]
Querying redis with pmseries directly e.g. `pmseries -p 6379 hinv.ncpu` works as expected, but querying redis via pmproxy with e.g. `pmseries -Dseries -p 44322 hinv.ncpu`  trips an assert in pmproxy :

```
$ pmseries -D series -p 44322 hinv.ncpu
Slot range: 00000-16384
    Master: localhost:44322
Redis [slot=00000] CLUSTER (null)
Redis [slot=00000] localhost:44322
Connected to Redis on localhost:44322
Redis [slot=12804] GET pcp:version:schema
Redis [slot=12804] localhost:44322
Redis [slot=12804] SET pcp:version:schema
Redis [slot=12804] localhost:44322
Disconnecting from localhost:44322 failed - Server closed the connection
```

In pmproxy.log -

`pmproxy: redis.c:1580: redisProcessCallBacks: Assertion `(c->flags & REDIS_SUBSCRIBED || c->flags & REDIS_MONITORING)' failed.`

```
1552         /*
1553          * Even if the context is subscribed, pending regular callbacks will
1554          * get a reply before pub/sub messages arrive.
1555          */
1556         if (__redisShiftCallBack(&ac->replies, &cb) != REDIS_OK) {
1557             /*
1558              * A spontaneous reply in a not-subscribed context can be the error
1559              * reply that is sent when a new connection exceeds the maximum
1560              * number of allowed connections on the server side.
1561              *
1562              * This is seen as an error instead of a regular reply because the
1563              * server closes the connection after sending it.
1564              *
1565              * To prevent the error from being overwritten by an EOF error the
1566              * connection is closed here. See issue #43.
1567              *
1568              * Another possibility is that the server is loading its dataset.
1569              * In this case we also want to close the connection, and have the
1570              * user wait until the server is ready to take our request.
1571              */
1572             if (((redisReply *)reply)->type == REDIS_REPLY_ERROR) {
1573                 c->err = REDIS_ERR_OTHER;
1574                 pmsprintf(c->errstr, sizeof(c->errstr), "%s", ((redisReply *)reply)->str);
1575                 c->reader->fn->freeObject(reply);
1576                 __redisAsyncDisconnect(ac);
1577                 return;
1578             }
1579             /* No more regular callbacks and no errors, the context *must* be subscribed or monitoring. */
1580             assert((c->flags & REDIS_SUBSCRIBED || c->flags & REDIS_MONITORING));
1581             if (c->flags & REDIS_SUBSCRIBED)
1582                 __redisGetSubscribeCallBack(ac, reply, &cb);
1583         }
```

gdb shows that c->flags == 0x02, which is REDIS_CONNECTED. So it seems pmproxy thinks it's connected but redis has sent an unexpected or async reply and disconnected the client.

Nathan could this be due to exceeding the maximum number of connections (as it says in the comment), or something else? Maybe something to do with the unexpected DISCONNECTED issue we're sometimes seeing when switching hosts in the Grafana pcp-redis overview dashboard?

**Comment by natoscott** [comment]
@goodwinos what is the reply type if not an error?  (line 1572)

**Comment by natoscott** [comment]
Looks like its a STRING with value "2" - the response from the schema version check, perhaps that's interfering with the protocol proxying somehow?

@goodwinos do you want me to debug this or are you planning to?

**Comment by goodwinos** [comment]
```
(gdb) p ((redisReply *)reply)->type
$2 = REDIS_REPLY_STRING
(gdb) p *((redisReply *)reply)
$3 = {type = REDIS_REPLY_STRING, integer = 0, len = 1, str = 0x55effc8f80e0 "2", elements = 0, element = 0x0}
(gdb) p *c
$4 = {err = 0, errstr = '\000' <repeats 127 times>, fd = 23, flags = 2, obuf = 0x55effc435e23 "", reader = 0x55effbe22800, 
  connection_type = REDIS_CONN_TCP, timeout = 0x0, tcp = {host = 0x55effbe22a60 "localhost", source_addr = 0x0, port = 6379}, 
  unix_sock = {path = 0x0}, saddr = 0x55effbe245d0, addrlen = 16, ssl = 0x0}

```

**Comment by goodwinos** [comment]
Resolved by 5d4556a268a48cf0b8ac8d9f6e0b801a5a027c76
